### ISSUE TITLE
400 Bad Request when adding Organization Plugin Config parameters/secrets.

### ISSUE THREAD

**Issue by ashcorpdev** [intelowlproject/IntelOwl#1821]
## What happened

When attempting to save a plugin configuration (any value in parameters or secrets) for an organization, the API responds with a 400 status code and returns an error message (See error message below).

## Environment
1. OS: Ubuntu 22.04
2. IntelOwl version: v5.0.1 (Docker)

## What did you expect to happen

I expected the parameters/secrets to be saved correctly to the database.

## How to reproduce your issue

On a fresh installation of IntelOwl v5.0.x, create a new organization and attempt to save a plugin configuration for it with any parameter/secret as the organization owner. (In my case, I used `Secrets > Connector > OpenCTI > url_key_name` and used `http://example.com`.)

### 400 Bad Request 

**Error message**:
`{"errors":{"organization":["Incorrect type. Expected pk value, received str."]}}`

**Request Payload from frontend**:

```json
{
  "create": true,
  "edit": true,
  "type": "2",
  "plugin_name": "OpenCTI",
  "attribute": "url_key_name",
  "value": "http://example.com",
  "organization": "My Org",
  "config_type": "2"
}
```

## Suspected cause

***Note**: I have very little experience with React and no Django experience, so I may be incorrect in my assumptions here. However, I believe I have identified the root cause of my problem.*

----

Since the 5.0.0 release, it appears a migration was applied to the database, which removed following fields from Database model:

- `organization`
- `attribute`
- `plugin_name`
- `type`
- `config_type`

*See [here](https://github.com/intelowlproject/IntelOwl/blob/4fb16c9f0c734c538e45e8dfe42444205ef8d6a7/api_app/migrations/0028_plugin_config.py#L138C1-L145C11).*

Looking at the included `PluginConfig` model, it seems there are only these fields:

- `owner`
- `for_organization`
- `parameter`

*See [here](https://github.com/intelowlproject/IntelOwl/blob/7916bc7ce32f8863b10f7ba7350bda734cbbe783/api_app/models.py#L474C1-L489C10)*.

According to the Postgres database on my installation, I have the following structure for the `api_app_pluginconfig` table:

- `id`
- `value`
- `updated_at`
- `owner_id`
- `parameter_id`
- `for_organization`

This appears to match the PluginConfig model with the new parameters.

However, [the **serializer** for the PluginConfig seems to still use the old field values](https://github.com/intelowlproject/IntelOwl/blob/7916bc7ce32f8863b10f7ba7350bda734cbbe783/api_app/serializers.py#L840C1-L852C10).

When adding a secret or parameter in a user plugin config, I **do not** receive the above error and the request is processed by the API as expected. This is likely because the `owner` field that is used in the serializer has not changed between updates (and the organization would be `None`, because there is no requirement for it).

```python
    owner = rfs.HiddenField(default=rfs.CurrentUserDefault())
```

The organization variable, however, has; it now uses the `PrimaryKeyRelatedField` serializer method instead of the `SlugRelatedField` method that was used in v4.2.3.

*See [here](https://github.com/intelowlproject/IntelOwl/blob/7916bc7ce32f8863b10f7ba7350bda734cbbe783/api_app/serializers.py#L891C1-L893C6).*

The error message appears to originate from this serializer method.

*See [here](https://github.com/encode/django-rest-framework/blob/3f8ab538c1a7e6f887af9fec41847e2d67ff674f/rest_framework/relations.py#L241)*.

I suspect that, because the field names do not match between model and serializer, the new serializer method may be trying to match with the wrong value against one (or all) of the fields, which won't be the expected type (as it is a string, not a primary key value).

I believe that the serializer needs to be updated to reflect the new database fields, but I do not know if that would (on its own) solve the issue.

**Comment by mlodic** [comment]
Thank you very much for the very detailed report. Super useful. We'll analyze this soon and come back with an answer

**Comment by 0ssigeno** [comment]
First of all, thank you for the detailed report.
The issue is easily reproducible on my side and the patch is currently under development in the `fix_plugin_config_org` branch.

  
The missmatch that you see in the serializer between the database current model and the fields, is because the backend and frontend follow two different development cycle, and the backend changes should not have caused issue for the frontend.
**Should** is the correct word.

The patch in the branch fix the error (on what I was able to test), and I have added some tests for it that apparently we did not have :(
